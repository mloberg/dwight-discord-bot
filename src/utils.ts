export function between(min: number, max: number): number {
    return Math.floor(Math.random()*(max-min+1)+min);
}

export function env(key: string, _default: string): string {
    return process.env[key] || _default;
}

export function rand<T>(items: T[]): T {
    return items[Math.floor(Math.random() * items.length)];
}

export function ucfirst(input: string): string {
    return input.charAt(0).toUpperCase() + input.slice(1).toLowerCase();
}

export function roll(expr: string): number {
    const input = expr.toLowerCase().match(/^(\d+)?d(\d+)([+-×x])?(\d+)?/);
    if (!input) {
        throw new Error("Invalid roll expression");
    }

    const count = Number(input[1] || 1);
    const dice = Number(input[2]);
    const modifier = input[3];
    const mod = Number(input[4]);

    let total = 0;
    for (let index = 0; index < count; index++) {
        total += Math.floor(Math.random() * (dice - 1 + 1)) + 1;
    }

    if ("+" === modifier) {
        total += mod;
    } else if ("-" === modifier) {
        total -= mod;
    } else if ("x" === modifier || "×" === modifier) {
        total = total * mod;
    }

    return total;
}
