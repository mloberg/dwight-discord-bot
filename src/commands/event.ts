import { Message } from 'discord.js';

import { Command } from '../types';
import { random } from '../utils';

export const events = [
    'A door opens',
    'A fire starts',
    'A meteor shoots across the sky',
    'A monster appears',
    'A screech pierces the air',
    'A storm begins',
    'A strange star appears in the sky',
    'A strong gust of wind blows through',
    'A tremor shakes the ground',
    'Someone experiences déjà vu',
    'Someone gets angry',
    'Someone glimpses the future',
    'Someone has a sense of foreboding',
    'Someone has to go to the bathroom',
    'Something spills or falls to the ground',
    "Something isn't where it's supposed to be",
    'The lights go out',
    'The sun comes out',
    "There's a foul smell in the air",
    'Unexplained magic occurs',
];

const command: Command = {
    name: 'event',
    description: 'Trigger a random event',
    async run(message: Message): Promise<Message> {
        await message.delete();

        return message.channel.send(random(events));
    },
};

export default command;
