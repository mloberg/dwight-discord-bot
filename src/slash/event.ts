import { SlashCommand } from '../types';
import { roll } from '../utils';

export const events = [
    'A door opens',
    'A fire starts',
    'A meteor shoots across the sky',
    'A monster appears',
    'A screech pierces the air',
    'A storm begins',
    'A strange star appears in the sky',
    'A strong gust of wind blows through',
    'A tremor shakes the ground',
    'Someone experiences déjà vu',
    'Someone gets angry',
    'Someone glimpses the future',
    'Someone has a sense of foreboding',
    'Someone has to go to the bathroom',
    'Something spills or falls to the ground',
    "Something isn't where it's supposed to be",
    'The lights go out',
    'The sun comes out',
    "There's a foul smell in the air",
    'Unexplained magic occurs',
];

const event: SlashCommand = {
    name: 'event',
    description: 'Trigger a random event',
    options: [
        {
            name: 'roll',
            description: 'd20 roll',
            type: 'INTEGER',
        },
    ],
    async run(command) {
        const dice = command.options.getInteger('roll') || roll('d20');
        const event = events[dice - 1];

        await command.reply(event);
    },
};

export default event;
