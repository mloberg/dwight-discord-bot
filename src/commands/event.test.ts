import { CommandInteraction } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import event, { events } from './event';

jest.mock('discord.js', () => ({
    CommandInteraction: jest.fn().mockImplementation(() => ({
        reply: jest.fn(),
        options: {
            getInteger: jest.fn(),
        },
    })),
}));

describe('/event', () => {
    it('is a slash command', () => {
        expect(event).toMatchSnapshot();
    });

    it('returns a random event', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never));

        await event.handle(command);
        expect(events).toContainEqual(command.reply.mock.calls[0][0]);
    });

    it('returns an event for a dice roll', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getInteger.mockReturnValue(1);

        await event.handle(command);
        expect(command.reply).toHaveBeenCalledWith('A door opens');
    });
});
