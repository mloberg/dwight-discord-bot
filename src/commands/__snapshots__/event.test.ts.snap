// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/event is a slash command 1`] = `
Object {
  "config": Object {
    "default_permission": undefined,
    "description": "Trigger a random event",
    "name": "event",
    "options": Array [
      Object {
        "autocomplete": undefined,
        "choices": undefined,
        "description": "Value of a d20 roll.",
        "max_value": 20,
        "min_value": 1,
        "name": "roll",
        "required": false,
        "type": 4,
      },
    ],
  },
  "handle": [Function],
}
`;
